### ISSUE TITLE
mini.clue: dynamic width normalization

### ISSUE THREAD

**Issue by wroyca** [echasnovski/mini.nvim#855]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

clue

### Description

Hello o/

When displaying clues with varying line widths, the float width is determined by the widest line, which is far too aggressive. One solution I'd like to propose here would be to calculate a width suitable to comfortably display "most" lines while truncating the excessively wide ones.

Example (late night pseudo code, don't use):

```lua
H.buffer_get_width = function()
  if not H.is_valid_buf(H.state.buf_id) then return end

  local lines = vim.api.nvim_buf_get_lines(H.state.buf_id, 0, -1, false)
  local lines_count = #lines
  local lines_width_total = 0

  for _, l in ipairs(lines) do
    lines_width_total = lines_width_total + vim.fn.strdisplaywidth(l)
  end

  local lines_width_average = lines_width_total / lines_count

  table.sort(lines, function(a, b)
    return vim.fn.strdisplaywidth(a) < vim.fn.strdisplaywidth(b)
  end)

  local lines_width = 0
  local lines_width_cumulative = 0
  local lines_width_factor = 2
  
  -- (too tired, revisit properly with coffee)
  for i, l in ipairs(lines) do
    lines_width_cumulative = lines_width_cumulative + vim.fn.strdisplaywidth(l)
    if lines_width_cumulative <= lines_width_average * lines_width_factor then
      lines_width = i
    else
      break
    end
  end

  local res = 0
  for i = 1, lines_width do
    res = res + vim.fn.strdisplaywidth(lines[i])
  end

  print (res) -- debug
  return res
end
```

### Before:
![image](https://github.com/echasnovski/mini.nvim/assets/91024200/72d6ea64-c933-4ad1-ad28-60d738eb09d0)
![image](https://github.com/echasnovski/mini.nvim/assets/91024200/ea698556-bb39-4456-a555-dd01e19d3028)

### After:
![image](https://github.com/echasnovski/mini.nvim/assets/91024200/0786e759-29e3-4ffc-9636-5c2f03eb2a7f)

![image](https://github.com/echasnovski/mini.nvim/assets/91024200/cf4dc1a3-9b3f-4a92-8e70-e88834cfd636)

### And more:
![Screenshot from 2024-05-04 20-59-33](https://github.com/echasnovski/mini.nvim/assets/91024200/0dfb357f-0ff5-4685-bffc-96ba1965a19c)

**Comment by echasnovski** [comment]
Thanks for the suggestion!

There is already a functionality for users to tweak window config to their liking, as [`config.window.config` can be callable](https://github.com/echasnovski/mini.nvim/blob/fb4df45edec418443e4801719f1cb98e89697724/doc/mini-clue.txt#L556-L558).

The main it seems to be one or two really long lines which make automated width too large. Here is one approach at dynamic width that I find reasonable compromise:

```lua
local compute_dynamic_width = function(buf_id)
  local max_width = 0.4 * vim.o.columns
  local widths = vim.tbl_map(vim.fn.strdisplaywidth, vim.api.nvim_buf_get_lines(buf_id, 0, -1, false))
  table.sort(widths)
  for i = #widths, 1, -1 do
    if widths[i] <= max_width then return widths[i] end
  end

  return max_width
end

local win_config = function(buf_id) return { width = compute_dynamic_width(buf_id) } end

local miniclue = require('mini.clue')
miniclue.setup({
  -- Clues and triggers setup
  window = { config = win_config },
})
```

It has user-defined maximum width (as a share of current Neovim instance width) and picks the window width so as to fit the largest present line below that threshold (if there is one).

------

My initial reaction was to leave it as possible configuration, but I *think* I might change my mind about somehow adding this type of behavior directly in 'mini.clue' (maybe by allowing fractional width?).

I'll think about it.

**Comment by echasnovski** [comment]
Sorry, but I think leaving it for a manual configuration is a better solution in the long term. Maybe if/when there is a wiki with all the suggested custom solutions, they will be easier to find.

Besides, all those descriptions can be adjusted manually with [`set_mapping_desc()`](https://github.com/echasnovski/mini.nvim/blob/b0b435873375c4341f2ecbcb95e9e645abe0d679/doc/mini-clue.txt#L624) and the original culprit of those long built-in descriptions was solved before 0.10 made stable release.

Closing as not planned.
